audio-hal: Calibration extravaganza

kholk opened the pull request:
Just remember to remove that draft status from the github PR to sonyxperiadev after merging this... :)))

MarijnS95 committed:
cirrus_sony: Do not unpack unused function pointers from init config

These functions are not used; it's rather wasteful readability-wise to
retrieve them.

(The others are not used yet... but will be in future commits)

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

MarijnS95 committed:
cirrus_sony: Enable deep buffer playback mixer through paths

Edo has the cirrus AMPs on a different backend (SEN_MI2S_RX to be exact,
instead of PRI on seine). The mixer listed here is to enable playback
through the deep-buffer usecase, so we can simply enable that path to
get the right backend enabled regardless of the device based on its
mixer_paths.

At the same time "clarify" hardcoded PCM id `0` by reading it from the
configuration (potentially overridden by audio_platform_info.xml)
instead.

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

MarijnS95 committed:
cirrus_sony: Select usecase and enable backend through extn API

Enable silent audio playback (for calibration) through more verbose but
more specific APIs instead of manually poking around with the paths.

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

MarijnS95 committed:
cirrus_sony: BAD HACK: Wait until `primary_output` becomes available

This pointer strangely has a value immediately on Edo, but not on Seine.

MarijnS95 committed:
cirrus_sony: Use tabs instead of spaces

MarijnS95 committed:
cirrus_sony: Do not sum error codes with positive string-length

cirrus_format_mixer_name returns the number of formatted characters
(result of snprintf). Returning this positive number is not always a
good idea, like when check_error_state returns 22 as the length of a
mixer control, which is caught by the `==0` check resulting in a full
reset.
Secondly, a postive length of 22 together with an actual, expected
failure of -EINVAL which conveniently is -22 results in 0; no error at
all!!

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

kholk committed:
cirrus_sony: Redo PCM handling for calibration, add locking

There is no need (and it's wrong) to call pcm_start before
writing the data, as we can simply retrieve, prepare and
write the silence data to the PCM instead.

Also, while at it, add some locking around to make sure that
we don't race while setting module states, especially in
corner cases where we need to calibrate the speakers before
starting processing (very, very unlikely case that should
anyway never happen).

kholk committed:
cirrus_sony: Cleanup unused development leftovers

During development, some code was left there commented out.
Let's just remove it, as it serves no purpose, and it never will.

kholk committed:
cirrus_sony: Fixes for new firmware mixer names and order

The new firmware features more creativity, which means that the
mixer names have changed. Also, for some reason, when we've got
a dual amp setup, the DSP refuses to load the R-SPK firmware by
caprice if we load the L-SPK one before that, even though they
definitely have no real correlation between each other (or at
least not firmware-wise).

Even though there are *way* better ways to do this, time does
not allow me to, so this is a very fast workaround to get the
audio calibrations kicking.

MarijnS95 commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -615,28 +626,28 @@ static int cirrus_play_silence(int seconds) {
+    while ((!adev->primary_output || !adev->platform) && adev_retry) {
+        ALOGI("%s: Waiting for audio device...", __func__);
+        sleep(1);
+	adev_retry--;
```
Fix your indentation, yo!

MarijnS95 commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -445,7 +456,7 @@ static int cirrus_set_mixer_value_by_name(char* ctl_name, int value) {
 
     ctl_config = mixer_get_ctl_by_name(card_mixer, ctl_name);
     if (!ctl_config) {
-        ALOGE("%s: Cannot get mixer control %s", __func__, ctl_name);
+        ALOGD("%s: Cannot get mixer control %s", __func__, ctl_name);
```
We should find a better way to be clever about mono vs stereo, seeing these warnings obviously in the logs is vital in most other scenarios.

MarijnS95 commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -615,28 +626,28 @@ static int cirrus_play_silence(int seconds) {
     fp_enable_snd_device(adev, uc_info_rx->out_snd_device);
     fp_enable_audio_route(adev, uc_info_rx);
 
+    select_devices(adev, uc_info_rx->id);
```
Whoa, that's a massively complex function! In the end it supersedes the call to `fp_enable_snd_device` + `fp_enable_audio_route`, all we need then is `fp_platform_check_and_set_codec_backend_cfg` afaik.

Anyway, as mentioned above it is not available to `audio_extn` libs.

MarijnS95 commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -645,25 +656,15 @@ static int cirrus_play_silence(int seconds) {
-        ret = -EINVAL;
-        goto exit;
-    }
+    handle.pcm_rx = handle.pcm_tx = NULL;
```
Lmao

kholk committed:
cirrus_sony: Consolidate calibration process

On devices with multiple CS35L41 amps the calibration process is a bit
slower for.. reasons.
First of all, the DSP likes to just take time to get back alive after
the calibration run succeeding (reasons unknown: firmware is closed)
and there is no way to know when it precisely comes back to life.

Then, this task being now very much time consuming brings more issues
to the table, like the audio HAL trying to playback sounds from the
OS before the calibration run is done: this may actually interact with
the calibration silence and/or confuse the DSP and make it crash, which
is why we now need to carefully set and check the module state, meaning
that we are obliged to check and forbid audio playback if this software
is waiting for DSP calibration.

The result of this is that *at least* Sony Edo, featuring this kind of
dual-amp cofiguration, can now calibrate and switch to protection FW
without requiring a system reboot, even though it's going to be a bit
slow: this was ignored for the sake of safety and audio quality through
the internal speakers and it is very important to recalibrate this
when we have the opportunity to, as the TA-stored calibration cannot be
accounting for hardware aging while, with this, WE CAN.

kholk committed:
cirrus_sony: Loop around DSP state after calibration

Instead of waiting for a arbitrary amount of time in any case,
a better user experience may be (in some cases) achieved by
looping through the DSP state: if we succeed to change the DSP
firmware mixer (to load the new FW) before the max wait, that's
going to be... well... less time to lose.

This actually brings benefits depending on the specific device
and on the specific boot: for example, my PDX203 sometimes
takes as much as 5 seconds, sometimes just two, and this probably
depends on the boot order since it's entirely async, which gets
different depending on the device state before actually booting,
as the hardware will be in different states and some daemons
init time will be skewed;

For example:
- Bootloader version X will leave some HW in a different state
  than version Y; or
- The device is booting from a fastboot continue command; or
- The device is cold booting.

MarijnS95 reviewed (changes requested):
Guess why these are not in any header, but (some!) only exposed through `spkr_prot_init_config_val.fp_*`...
```c
ld.lld: error: undefined symbol: select_devices
>>> referenced by cirrus_sony.c:649 (vendor/qcom/opensource/audio-hal/primary-hal/hal/audio_extn/cirrus_sony.c:649)
>>>               out/target/product/pdx206/obj/SHARED_LIBRARIES/libcirrusspkrprot_intermediates/cirrus_sony.o:(cirrus_play_silence)

ld.lld: error: undefined symbol: pcm_open_prepare_helper
>>> referenced by cirrus_sony.c:662 (vendor/qcom/opensource/audio-hal/primary-hal/hal/audio_extn/cirrus_sony.c:662)
>>>               out/target/product/pdx206/obj/SHARED_LIBRARIES/libcirrusspkrprot_intermediates/cirrus_sony.o:(cirrus_play_silence)

ld.lld: error: undefined symbol: platform_invalidate_backend_config
>>> referenced by cirrus_sony.c:704 (vendor/qcom/opensource/audio-hal/primary-hal/hal/audio_extn/cirrus_sony.c:704)
>>>               out/target/product/pdx206/obj/SHARED_LIBRARIES/libcirrusspkrprot_intermediates/cirrus_sony.o:(cirrus_play_silence)
```

@kholk can you please check the "Allow edits from maintainers" so that I can push a rebased version of this PR? GH is lost and makes it a mess now, because I squashed+force-pushed some cleanups into my own commits (where they belong).

kholk commented:
> Guess why these are not in any header, but (some!) only exposed through `spkr_prot_init_config_val.fp_*`...
> 
> ```c
> ld.lld: error: undefined symbol: select_devices
> >>> referenced by cirrus_sony.c:649 (vendor/qcom/opensource/audio-hal/primary-hal/hal/audio_extn/cirrus_sony.c:649)
> >>>               out/target/product/pdx206/obj/SHARED_LIBRARIES/libcirrusspkrprot_intermediates/cirrus_sony.o:(cirrus_play_silence)
> 
> ld.lld: error: undefined symbol: pcm_open_prepare_helper
> >>> referenced by cirrus_sony.c:662 (vendor/qcom/opensource/audio-hal/primary-hal/hal/audio_extn/cirrus_sony.c:662)
> >>>               out/target/product/pdx206/obj/SHARED_LIBRARIES/libcirrusspkrprot_intermediates/cirrus_sony.o:(cirrus_play_silence)
> 
> ld.lld: error: undefined symbol: platform_invalidate_backend_config
> >>> referenced by cirrus_sony.c:704 (vendor/qcom/opensource/audio-hal/primary-hal/hal/audio_extn/cirrus_sony.c:704)
> >>>               out/target/product/pdx206/obj/SHARED_LIBRARIES/libcirrusspkrprot_intermediates/cirrus_sony.o:(cirrus_play_silence)
> ```
> 
> @kholk can you please check the "Allow edits from maintainers" so that I can push a rebased version of this PR? GH is lost and makes it a mess now, because I squashed+force-pushed some cleanups into my own commits (where they belong).

It's checked since year 0.... :P

MarijnS95 commented:
@kholk Except that it isn't? github says:

```
Add more commits by pushing to the generic-usecase branch on kholk/vendor-qcom-opensource-audio-hal-primary-hal.
```

But github also says:

```console
me:primary-hal/ $ gpf git@github.com:kholk/vendor-qcom-opensource-audio-hal-primary-hal.git HEAD:generic-usecase
ERROR: Permission to kholk/vendor-qcom-opensource-audio-hal-primary-hal.git denied to MarijnS95.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

MarijnS95 commented:
I even tried updating the PR directly:

```
me:audio-hal/ $ gpf origin HEAD:refs/pull/1/head
...
To github.com:MarijnS95/vendor-qcom-opensource-audio-hal-primary-hal.git
 ! [remote rejected]     HEAD -> refs/pull/1/head (deny updating a hidden ref)
error: failed to push some refs to 'github.com:MarijnS95/vendor-qcom-opensource-audio-hal-primary-hal.git'
```

kholk committed:
audio_extn: cirrus-sony: Add dependency on audio_primary

We are now using functions from audio_hw for cirrus.
